cilium/CFP-30235-xds-adapter.md

```diff
@@ -0,0 +1,204 @@
+3. All changes to endpoint routing currently have to go through Services or
+   EndpointSlices, and by extension, the Kubernetes API Server. It would be very
+   expensive to implement incremental weighting/spillover adjustments with that
+   overhead. For example, adjusting the weight of any individual endpoint would
+   require a write to the EndpointSlice API which would then need to be
+   distributed to all consumers of that API. That can be problematic from a
+   scalability perspective when there are frequent updates paired with a large
+   number of endpoints and/or nodes.
+
+I believe that xDS is uniquely positioned to address these limitations.
+[LRS](https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/load_stats/v3/lrs.proto)
+(Load Reporting Service) provides a straightforward way to provide load reports
+to a centralized control plane. Similarly, [Delta
+xDS](https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/xds_api#delta-endpoints)
+enables xDS to distribute incremental updates that can be smaller in scope than
+comparable Kubernetes APIs. All of this could be combined to avoid adding
+additional load on the Kubernetes API Server.
+
+#### Endpoints From Other Sources
+
+In some cases, it’s useful to be able to route to endpoints from other sources,
+such as other Kubernetes clusters or outside of Kubernetes altogether. This can
+be particularly relevant when implementing Multi-Cluster Service routing.
+Implementations today often rely on one of the following approaches:
+
+1. Mirror endpoints from other clusters into Kubernetes by creating custom
+   Services and EndpointSlices.
+2. Introduce Gateways on the edges of clusters that can be used to route
+   requests from other clusters.
+
+With xDS support, it could be possible to have a centralized xDS control plane
+that was familiar with endpoints in multiple sources, providing a potentially
+more efficient alternative to either of the approaches commonly used today.
+
+## Proposal
+
+### Logic
+
+The introduction of this adapter would mean that Cilium could accept endpoints
+from more than one source. Cilium could be configured to read exclusively from
+Kubernetes (default), xDS, or both, depending on the use case.
+
+When both data sources were in use, precedence in any conflicts would be given
+to data received directly from Kubernetes. As a concrete example, it's possible
+that the same Service could be represented in both Kubernetes and xDS. If that
+were the case, we'd give precedence to the Service from Kubernetes APIs and drop
+the one from xDS, likely surfacing some kind of warning along the way.
+
+### Architecture
+
+This xDS adapter would be similar conceptually to KVStore. Essentially this
+would be another backend that Cilium could get data from. Similar to how Cilium
+can read from Kubernetes APIs and/or KVStore today, we would add a third option
+for xDS as a data source, and develop a shared interface that worked across all
+data sources.
```

Comment:
Worth sharing in more detail what this would entail in terms of refactoring the agent functionality to support xDS as an alternative store.
Further, since LRS enables the client to share load information with the server, how would this information be surfaced and interact with rest of the agent functionality, given this capability is unavailable in the agent today?

Reply:
I tried to address this in the latest update, adding a new section for LRS, but let me know if I missed anything.